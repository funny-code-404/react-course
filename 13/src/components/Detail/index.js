import React from "react";
import PropTypes from "prop-types";

import "./detail.css";
import DetailCreatedIn from "./CreatedIn";
import DetailUniqueUnit from "./UniqueUnit";
import DetailUniqueTech from "./UniqueTech";
import CivilizationBonus from "./CivilizationBonus";
import AttackBonus from "./AttackBonus";
import Special from "./Special";
import Cost from "./Cost";

class Detail extends React.Component {
    render() {
        const {
            name,
            army_type,
            expansion,
            team_bonus,
            civilization_bonus,
            unique_unit,
            unique_tech,
            description,
            age,
            created_in,
            cost,
            build_time,
            reload_time,
            attack_delay,
            movement_rate,
            line_of_sight,
            hit_points,
            range,
            attack,
            armor,
            accuracy,
            special,
            attack_bonus,
            search_radius,
        } = this.props.data;

        const arrs = [
            { army_type },
            { team_bonus },
            { description },
            { expansion },
            { age },
            { build_time },
            { reload_time },
            { attack_delay },
            { movement_rate },
            { line_of_sight },
            { hit_points },
            { range },
            { attack },
            { armor },
            { search_radius },
            { accuracy },
        ];

        return (
            <div className="detail">
                <h1>{name}</h1>
                <div className="ditail-block">
                    {arrs.map((x) => {
                        for (const key in x) {
                            return (
                                x[key] && (
                                    <p key={key}>
                                        {key}: <span>{x[key]}</span>
                                    </p>
                                )
                            );
                        }
                    })}
                    {cost && <Cost data={cost} />}
                    {attack_bonus && <AttackBonus data={attack_bonus} />}
                    {special && <Special data={special} />}
                    {civilization_bonus && (
                        <CivilizationBonus data={civilization_bonus} />
                    )}
                    {unique_unit && Boolean(unique_unit.length) && (
                        <DetailUniqueUnit data={unique_unit} />
                    )}
                    {unique_tech && Boolean(unique_tech.length) && (
                        <DetailUniqueTech data={unique_tech} />
                    )}
                    {created_in && Boolean(created_in.length) && (
                        <DetailCreatedIn data={created_in} />
                    )}
                </div>
            </div>
        );
    }
}

Detail.propTypes = {
    data: PropTypes.objectOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
            PropTypes.arrayOf(PropTypes.string),
            PropTypes.objectOf(
                PropTypes.oneOfType([PropTypes.number, PropTypes.string])
            ),
        ])
    ),
};

export default Detail;
