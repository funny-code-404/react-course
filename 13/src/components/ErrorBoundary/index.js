import React from "react";
import PropTypes from 'prop-types'

class ErrorBoundary extends React.Component {
    state = {
        error: false,
    }

    componentDidCatch(e) {
        if(e){
            this.setState({ error: true })
        }
    }

    render(){
        const { error } =this.state;
        return error ? <h1>Warning, you got the error</h1> : this.props.children
    }
}

ErrorBoundary.propTypes = {
    children: PropTypes.node.isRequired,
};

export default ErrorBoundary;