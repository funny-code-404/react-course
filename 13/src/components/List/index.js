import { Link } from "react-router-dom";
import PropTypes from "prop-types";

import './list.css'

const List = ({data, title, path}) => {
	return  (
		<div className='list'>
			<h1>{ title }</h1>
			<ul>
				{
					Boolean(data.length) &&
						data.map(e => (
							<li key={e.id} className='li'>
								<Link to={`${path}/${e.id}`} className='link'>{e.name}</Link>
							</li>
					))
				}
			</ul>
		</div>
	)
}

List.propTypes = {
	data: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.oneOfType([
		PropTypes.number,
		PropTypes.string,
		PropTypes.arrayOf(PropTypes.oneOfType([
			PropTypes.number,
			PropTypes.string,
		])),
		PropTypes.objectOf(PropTypes.oneOfType([
			PropTypes.number,
			PropTypes.string,
			PropTypes.objectOf(PropTypes.oneOfType([
				PropTypes.number,
				PropTypes.string,
			]))
		]))
	]))),
	title: PropTypes.string.isRequired,
	path: PropTypes.string.isRequired,
}

export default List;