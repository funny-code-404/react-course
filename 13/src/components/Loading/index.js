import React from "react";
import { connect } from "react-redux";

import './loading.css';

export class Loading extends React.Component {
    get isLoading() {
        const {
            civilizations, units, structures,technologies,
            item,createdIn,uniqueTech,uniqueUnit, auth
        } = this.props;

        return civilizations.isFetching || units.isFetching || structures.isFetching || technologies.isFetching || 
                item.isFetching || createdIn.isFetching || uniqueTech.isFetching || uniqueUnit.isFetching || auth.isFetching
    }
    render(){
        return(
            this.isLoading &&
            <div className='loadin'>
                <div className="loader">
                    <div className="inner one"></div>
                    <div className="inner two"></div>
                    <div className="inner three"></div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => state

export default connect(mapStateToProps)(Loading);