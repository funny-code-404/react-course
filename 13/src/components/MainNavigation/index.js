import React from 'react';
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import { getData } from "../../redux/action";

export class MainNavigation extends React.Component {
	handleClick = () => {
		const { path } = this.props;
		this.props.getDataAction(path)
	}

	render(){
		return(
			<li>
				<Link to={this.props.path} onClick={this.handleClick} >{ this.props.title }</Link>
			</li>
		)
	}
}

const mapStateToProps = state => state;


const mapDispatchToProps = dispatch => ({
	getDataAction: path => dispatch(getData(path))
})

MainNavigation.propTypes = {
	path: PropTypes.string.isRequired,
	title: PropTypes.string.isRequired,
}

export default connect(mapStateToProps, mapDispatchToProps)(MainNavigation)