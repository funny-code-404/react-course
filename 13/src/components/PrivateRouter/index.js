import { connect } from "react-redux";
import { Route, Redirect } from "react-router-dom";

const PrivateRoute = ({ auth, component: Component, ...rest }) => (
	<Route { ...rest } render={ props => (
		auth.user  ? <Component {...props} /> : <Redirect to={{ pathname: '/login'}} />
	)} />
)

const mapStateToProps = state => state;

export default connect(mapStateToProps)(PrivateRoute);