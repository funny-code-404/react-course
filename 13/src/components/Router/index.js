import { Switch, Route } from "react-router-dom";

import { Civilizations, Home, Units, Structures, Technologies, Item, NotFound, Login, SingUp } from "../../pages";
import PrivateRoute from "../PrivateRouter";

const Routers = () => (
	<Switch>
		<PrivateRoute path='/' exact component={Home} />
		<Route path='/login' exact component={Login} />
		<Route path='/signup' exact component={SingUp} />
		<PrivateRoute path='/civilizations' exact component={Civilizations} />
		<PrivateRoute path='/units' exact component={Units} />
		<PrivateRoute path='/structures' exact component={Structures} />
		<PrivateRoute path='/technologies' exact component={Technologies} />

		<PrivateRoute path='/civilizations/:id' exact component={Item} />
		<PrivateRoute path='/units/:id' exact component={Item} />
		<PrivateRoute path='/structures/:id' exact component={Item} />
		<PrivateRoute path='/technologies/:id' exact component={Item} />
		<PrivateRoute component={NotFound} />
	</Switch>
)

export default Routers;