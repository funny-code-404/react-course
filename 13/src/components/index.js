export { default as Routers } from './Router';
export { default as MainNavigation } from './MainNavigation';
export { default as List } from './List';
export { default as Detail } from './Detail';
export { default as ErrorBoundary } from './ErrorBoundary';
export { default as Loading } from './Loading';
export { default as SignOut } from './signOut';