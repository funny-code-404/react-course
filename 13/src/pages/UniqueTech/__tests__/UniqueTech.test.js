import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import { UniqueTech } from "..";
import { Detail } from "../../../components";

configure({ adapter: new Adapter() });

describe('PAGES  UNIQUE TECH', () => {
    const mockFn = jest.fn();

    const props = {
        uniqueTechData: mockFn,
        uniqueTech: { data: 'text...'},
        url: '/path/1',
    }

    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<UniqueTech {...props} />);
    });

    it('to Match Snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
    it('component is defined', () => {
        expect(wrapper).toHaveLength(1)
    })
    it('Component is return 1 content', () => {
        expect(wrapper.find(Detail)).toHaveLength(1);
    });
     it('function should be called in componentDidMount', () => {
        expect(mockFn).toHaveBeenCalledTimes(1)
    })
});

// npm run test src/pages/UniqueTech