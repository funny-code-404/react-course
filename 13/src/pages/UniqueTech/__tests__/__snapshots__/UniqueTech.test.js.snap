// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PAGES  UNIQUE TECH to Match Snapshot 1`] = `
<Detail
  data="text..."
/>
`;
